src/eval.py

```diff
@@ -231,44 +247,156 @@ def partitioned_metrics(
+        hyperparams = Hyperparams()
+        model = self._construct_finetuning_model(pretrained_model)
+
+        parameters = param_groups_lrd(model)
+        optimizer = AdamW(parameters, lr=hyperparams.lr)
+
+        train_ds = WorldCerealLabelledDataset(self.train_df)
+        val_ds = WorldCerealLabelledDataset(self.val_df)
+
+        pos = (train_ds.df.LANDCOVER_LABEL == 11).sum()
+        wts = 1 / torch.tensor([len(train_ds.df) - pos, pos])
+        loss_fn = nn.BCEWithLogitsLoss(pos_weight=(wts / wts[0])[1])
+
+        generator = torch.Generator()
+        generator.manual_seed(self.seed)
+        train_dl = DataLoader(
+            train_ds,
+            batch_size=hyperparams.batch_size,
+            shuffle=True,
+            num_workers=hyperparams.num_workers,
+            generator=generator,
+        )
+
+        val_dl = DataLoader(
+            val_ds,
+            batch_size=hyperparams.batch_size,
+            shuffle=False,
+            num_workers=hyperparams.num_workers,
+        )
+
+        train_loss = []
+        val_loss = []
+        best_loss = None
+        best_model_dict = None
+        epochs_since_improvement = 0
+
+        for _ in tqdm(range(hyperparams.max_epochs), desc="Finetuning"):
+            model.train()
+            epoch_train_loss = 0.0
+            for x, y, dw, latlons, month, _, variable_mask in tqdm(
+                train_dl, desc="Training", leave=False
+            ):
+                x, y, dw, latlons, month, variable_mask = [
+                    t.to(device) for t in (x, y, dw, latlons, month, variable_mask)
+                ]
+                optimizer.zero_grad()
+                preds = model(
+                    x,
+                    dynamic_world=dw.long(),
+                    mask=variable_mask,
+                    latlons=latlons,
+                    month=month,
+                )
+                loss = loss_fn(preds.squeeze(-1), y.float())
+                epoch_train_loss += loss.item()
+                loss.backward()
+                optimizer.step()
+            train_loss.append(epoch_train_loss / len(train_dl))
+
+            model.eval()
+            all_preds, all_y = [], []
+            for x, y, dw, latlons, month, _, variable_mask in val_dl:
+                x, y, dw, latlons, month, variable_mask = [
+                    t.to(device) for t in (x, y, dw, latlons, month, variable_mask)
+                ]
+                with torch.no_grad():
+                    preds = model(
+                        x,
+                        dynamic_world=dw.long(),
+                        mask=variable_mask,
+                        latlons=latlons,
+                        month=month,
+                    )
+                    all_preds.append(preds.squeeze(-1))
+                    all_y.append(y.float())
+
+            val_loss.append(loss_fn(torch.cat(all_preds), torch.cat(all_y)))
+
+            try:
+                import wandb
```

Comment:
nit - can we have this not in a loop? (so that it doesn't try importing every time, if its not present)

maybe something like (early in the function):

```python
run = None
try:
    import wandb
    ... 
    run = wandb.run 
```

and then here:
```python
if run is not None:
    ....
```

Reply:
Yes that's better! :)